### PR TITLE
Fix problem with old Theano batch normalization and GPU

### PR DESCRIPTION
**Pull request by gvtulder**
This might solve #4881. See also Theano/Theano#5385.

There seems to be a problem in the gradient Theano's `IfElse` that is triggered by giving it GPU variables as input. In the example script in issue #4881, this happens because two batch normalization layers are stacked. Directly calling the Theano functions from `theano.sandbox.cuda.dnn` moves the variables to the gpu, which triggers the problem for the second BN layer.

As a workaround, this patch avoids the problem by converting the variables back to normal non-gpu variables before they are passed to the `IfElse`.

This is only a problem in the current batch normalization implementation, since the new version does not call the gpu functions directly.

### CODE REVIEW

**Comment by fchollet** [comment]
Could this cause performance issues?

**Comment by gvtulder** [comment]
No, I don't think it will cause performance issues.

There would be a performance issue if you previously had something like `conv_on_gpu(IfElse(batchnorm_on_gpu))`, which this patch changes to `conv_on_gpu(GpuFromHost(IfElse(HostFromGpu(batchnorm_on_gpu)))`. However, the Theano optimizations should notice that and should remove the unnecessary `GpuFromHost` and `HostFromGpu`.

I checked this for the script in #4881. The compiled result looks good: the variables before and after the IfElse stay on the GPU, so there should be no difference in performance.

**Comment by fchollet** [comment]
Thanks, LGTM

**Comment by naomifridman** [comment]
How can I install this correction ?
I use anaconda python, with the following packages:
keras        1.2.0                    py27_0    jaikumarm
theano      0.9.0.dev4            py27_0    jaikumarm

**Comment by Danielhiversen** [comment]
Upgrade to Keras 2
